UnicornRender/video/include/unicorn/video/Camera.hpp

```diff
@@ -8,141 +8,172 @@
+    UNICORN_EXPORT glm::vec3 const& GetUpVector() const;
+
+    /** @brief Returns position of camera */
+    UNICORN_EXPORT glm::vec3 const& GetPosition() const;
+
+    /** @brief Checks if camera view matrix must be updated and do it */
+    UNICORN_EXPORT void Frame();
+protected:
+    CameraController(glm::mat4& cameraView);
+
+    /** @brief Updates view matrix */
+    void UpdateViewMatrix();
+
+    glm::mat4& m_cameraView;
+    glm::vec3 m_position, m_upVector, m_direction, m_rightVector;
+    float speed;
+
+    bool m_isDirty;
+};
+
+/** @brief Camera projection abstract class for orthographical and perspetive projections */
+class CameraProjection
+{
+public:
+    /** @brief Checks if camera projection matrix must be updated and do it */
+    UNICORN_EXPORT void Frame();
+protected:
+    CameraProjection(system::Window* window, glm::mat4& cameraProj);
+    virtual ~CameraProjection();
+
+    /** @brief Disconnects from windows signals */
+    void Disconnect(system::Window*);
+
+    /** @brief Recalculates aspect ration on window size changed signal */
+    void OnWindowSizeChanged(system::Window*, std::pair<int32_t, int32_t> windowSize);
+
+    /** @brief Wheel scroll action */
+    virtual void Scroll(float yoffset) = 0;
+    virtual void UpdateProjection() = 0;
+
+    float m_aspect;
+    glm::mat4& m_cameraProjection;
+    system::Window* m_pWindow;
+    bool m_isDirty;
+};
+
+/** @brief Perspective camera projection controller */
+class PerspectiveCamera : public CameraProjection
+{
+public:
+    UNICORN_EXPORT PerspectiveCamera(system::Window* window, glm::mat4& cameraProj);
+
+    /** @brief Changes fov */
+    UNICORN_EXPORT void Scroll(float yoffset) override;
+
+    /** @brief Sets fov */
     UNICORN_EXPORT void SetFov(float fov);
 
-    /**
-     * @brief Returns direction
-     * @return Direction vector
-     */
-    UNICORN_EXPORT const glm::vec3& GetDirection() const;
-
-    /**
-     * @brief Returns up vector
-     * @return Up vector value
-     */
-    UNICORN_EXPORT const glm::vec3& GetUpVector() const;
-
-    /**
-     * @brief Returns projection matrix
-     * @return Projection matrix value
-     */
-    UNICORN_EXPORT const glm::mat4& GetProjection() const;
-
-    /**
-     * @brief Returns view matrix
-     * @return View matrix value
-     */
-    UNICORN_EXPORT const glm::mat4& GetView() const;
-
-    /**
-     * @brief Returns field of view value
-     * @return Field of view value
-     */
+    /** @brief Sets near Z plane */
+    UNICORN_EXPORT void SetZNear(float znear);
+
+    /** @brief Sets far Z plane */
+    UNICORN_EXPORT void SetZFar(float zfar);
+
+    /** @brief Returns fov */
     UNICORN_EXPORT float GetFov() const;
+private:
+    float m_fov, m_fovLowerBound, m_fovUpperBound, m_znear, m_zfar;
 
-    /**
-    * @brief Checks if view or projection need to be updated and update it, if needed
-    */
-    UNICORN_EXPORT void Frame();
+    void UpdateProjection() override final;
```

Comment:
Maybe use `final` in the class declaration, not in methods declaration, if at all? Unless planning to inherit from it and partially redefine methods.